src/txn/txn_rollback_to_stable.c

```diff
@@ -0,0 +1,396 @@
+	 */
+	if (ret != 0 || cval.len == 0)
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable "
+		    "requires a timestamp in the configuration string");
+	WT_RET(__wt_txn_parse_timestamp(session,
+	    "rollback_to_stable", &rollback_timestamp, &cval));
+
+	/*
+	 * Ensure the eviction server is out of the file - we don't
+	 * want it messing with us. This step shouldn't be required, but
+	 * it simplifies some of the reasoning about what state trees can
+	 * be in.
+	 */
+	WT_RET(__wt_evict_file_exclusive_on(session));
+	ret = __txn_rollback_to_stable_btree_walk(
+	    session, &rollback_timestamp);
+	__wt_evict_file_exclusive_off(session);
+
+	return (ret);
+}
+
+/*
+ * __txn_rollback_to_stable_check --
+ *	Ensure the rollback request is reasonable.
+ */
+static int
+__txn_rollback_to_stable_check(
+    WT_SESSION_IMPL *session, wt_timestamp_t *rollback_timestamp)
+{
+	WT_TXN_GLOBAL *txn_global;
+	WT_TXN_STATE *s;
+	uint32_t i, session_cnt;
+	int cmp;
+
+	txn_global = &S2C(session)->txn_global;
+	__wt_readlock(session, &txn_global->rwlock);
+	cmp = __wt_timestamp_cmp(
+	    rollback_timestamp, &txn_global->stable_timestamp);
+	__wt_readunlock(session, &txn_global->rwlock);
+	if (cmp < 0)
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable requires a "
+		    "timestamp greater or equal to the stable timestamp");
+
+	/*
+	 * Help the user - see if they have any active transactions. I'd
+	 * like to check the transaction running flag, but that would
+	 * require peeking into all open sessions, which isn't really
+	 * kosher.
+	 */
+	WT_ORDERED_READ(session_cnt, S2C(session)->session_cnt);
+	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++) {
+		if (s->id != WT_TXN_NONE || s->pinned_id != WT_TXN_NONE)
+			WT_RET_MSG(session, EINVAL,
+			    "rollback_to_stable not supported with "
+			    "active transactions");
+	}
+
+	return (0);
+}
+#endif
+
+/*
+ * __wt_txn_rollback_to_stable --
+ *	Rollback all in-memory state related to timestamps more recent than
+ *	the passed in timestamp.
+ */
+int
+__wt_txn_rollback_to_stable(
+    WT_SESSION_IMPL *session, const char *cfg[])
+{
+#ifndef HAVE_TIMESTAMPS
+	WT_UNUSED(cfg);
+
+	WT_RET_MSG(session, EINVAL, "rollback_to_stable "
+	    "requires a version of WiredTiger built with timestamp "
+	    "support");
+#else
+	WT_CONNECTION_IMPL *conn;
+	WT_CONFIG_ITEM cval;
+	WT_DECL_RET;
+	WT_DECL_TIMESTAMP(rollback_timestamp)
+
+	conn = S2C(session);
+	/*
+	 * Get the timestamp.
+	 */
+	ret = __wt_config_gets(session, cfg, "timestamp", &cval);
+	if (ret != 0 || cval.len == 0)
+		WT_RET_MSG(session, EINVAL, "rollback_to_stable "
+		    "requires a timestamp in the configuration string");
+
+	WT_RET(__wt_txn_parse_timestamp(session,
+	    "rollback_to_stable", &rollback_timestamp, &cval));
+	WT_RET(__txn_rollback_to_stable_check(
+	    session, &rollback_timestamp));
+
+	/* Allocate a non-durable btree bitstring */
+	WT_RET(__bit_alloc(session,
+	    conn->next_file_id, &conn->stable_rollback_bitstring));
```

Comment:
FTR, if an application cycled through 100 million files, this would be 12.8MB of memory.  If an application was creating a table every second, it could run for >3 years before getting to this point.